File: 2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level.js

    function minSteps(level) {
        const n = level.length;
        const eleInd = level.map((val, i) => [val, i]);
        eleInd.sort((a, b) => a[0] - b[0]);

        let vis = Array(n).fill(false);
        let ans = 0;

        for (let i = 0; i < n; i++) {
            if (vis[i] || eleInd[i][1] === i) continue;

            let cycle = 0, j = i;
            while (!vis[j]) {
                vis[j] = true;
                j = eleInd[j][1];
                cycle++;
            }
            ans += cycle - 1;
        }

        return ans;
    }

    function minimumOperations(root) {
        if (!root) return 0;

        let ans = 0;
        let q = [root];

        while (q.length) {
            const level = [];
            const nextLevel = [];

            for (const node of q) {
                level.push(node.val);
                if (node.left) nextLevel.push(node.left);
                if (node.right) nextLevel.push(node.right);
            }

            q = nextLevel;
            ans += minSteps(level);
        }

        return ans;
    };

File: 2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level.java
import java.util.*;

class Solution {
    private int minSteps(List<Integer> level) {
        int n = level.size();
        List<int[]> eleInd = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            eleInd.add(new int[]{level.get(i), i});
        }

        eleInd.sort(Comparator.comparingInt(a -> a[0]));
        boolean[] vis = new boolean[n];
        int ans = 0;

        for (int i = 0; i < n; i++) {
            if (vis[i] || eleInd.get(i)[1] == i) continue;

            int cycle = 0, j = i;
            while (!vis[j]) {
                vis[j] = true;
                j = eleInd.get(j)[1];
                cycle++;
            }
            ans += cycle - 1;
        }

        return ans;
    }

    public int minimumOperations(TreeNode root) {
        if (root == null) return 0;

        int ans = 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int sz = q.size();
            List<Integer> level = new ArrayList<>();

            while (sz-- > 0) {
                TreeNode node = q.poll();
                level.add(node.val);

                if (node.left != null) q.add(node.left);
                if (node.right != null) q.add(node.right);
            }

            ans += minSteps(level);
        }

        return ans;
    }
}
